src/controllers/UserLoginController.ts

```diff
@@ -1,60 +1,176 @@
 import { ControllerBase } from './ControllerBase';
-import type { LoginCredentials as DheLoginCredentials } from '@deephaven-enterprise/jsapi-types';
-import { REQUEST_DHE_USER_CREDENTIALS_CMD } from '../common';
+import {
+  CREATE_AUTHENTICATED_CLIENT_CMD,
+  GENERATE_DHE_KEY_PAIR_CMD,
+} from '../common';
+import { Logger, runUserLoginWorkflow } from '../util';
+import type {
+  IAsyncCacheService,
+  IDheClientFactory,
+  ISecretService,
+  IToastService,
+  ServerState,
+} from '../types';
+import { hasInteractivePermission } from '../dh/dhe';
+
+const logger = new Logger('UserLoginController');
 
 /**
  * Controller for user login.
  */
 export class UserLoginController extends ControllerBase {
-  constructor(dheCredentialsCache: URLMap<DheLoginCredentials>) {
+  constructor(
+    dheClientCache: URLMap<DheAuthenticatedClient>,
+    dheClientFactory: IDheClientFactory,
+    dheJsApiCache: IAsyncCacheService<URL, DheType>,
+    secretService: ISecretService,
+    toastService: IToastService
+  ) {
     super();
-    this.dheCredentialsCache = dheCredentialsCache;
+
+    this.dheClientCache = dheClientCache;
+    this.dheClientFactory = dheClientFactory;
+    this.dheJsApiCache = dheJsApiCache;
+    this.secretService = secretService;
+    this.toast = toastService;
 
     this.registerCommand(
-      REQUEST_DHE_USER_CREDENTIALS_CMD,
-      this.onDidRequestDheUserCredentials
+      GENERATE_DHE_KEY_PAIR_CMD,
+      this.onDidRequestGenerateDheKeyPair
+    );
+
+    this.registerCommand(
+      CREATE_AUTHENTICATED_CLIENT_CMD,
+      this.onCreateAuthenticatedClient
     );
   }
 
-  private readonly dheCredentialsCache: URLMap<DheLoginCredentials>;
+  private readonly dheClientCache: URLMap<DheAuthenticatedClient>;
+  private readonly dheClientFactory: IDheClientFactory;
+  private readonly dheJsApiCache: IAsyncCacheService<URL, DheType>;
+  private readonly secretService: ISecretService;
+  private readonly toast: IToastService;
 
   /**
-   * Handle the request for DHE user credentials. If credentials are provided,
-   * they will be stored in the credentials cache.
-   * @param serverUrl The server URL to request credentials for.
-   * @returns A promise that resolves when the credentials have been provided or declined.
+   * Handle request for generating a DHE key pair.
+   * @param serverState The server state to generate the key pair for.
    */
-  onDidRequestDheUserCredentials = async (serverUrl: URL): Promise<void> => {
-    // Remove any existing credentials for the server
-    this.dheCredentialsCache.delete(serverUrl);
+  onDidRequestGenerateDheKeyPair = async (
+    serverState: ServerState
+  ): Promise<void> => {
+    const serverUrl = serverState.url;
+
+    const title = 'Generate Private Key';
 
-    const username = await vscode.window.showInputBox({
-      placeHolder: 'Username',
-      prompt: 'Enter your Deephaven username',
+    const credentials = await runUserLoginWorkflow({
```

Comment:
One thing that's weird about this is that if we're already authenticated, it requests our username/password again. We should at least fill in the `lastLogin` here so we don't need to type in the user name again.
Ideally we'd use the same username, and just ask for the password again (like how other apps require re-authentication when performing security operations). It's somewhat odd having to enter the username again.

That being said, after using it - why do we have the user explicitly ask to generate a keypair? Is there any reason they'd want to create multiple key pairs? We're essentially generating this key pair so the user doesn't need to re-enter their password every time, correct? In which case, I'd expect the flow to be more like Login -> enter username/password, with an option to "Remember Me", rather than having an action "Generate DHE Keypair". Thoughts?

Reply:
Users could in theory create different keys for different usernames, so we would at least need to have a way for them to opt to login with another username / password. The remember me option is interesting depending on how much we want to expose to the user about the method of how things are stored. As-is, the user has to have some understanding they are opting for key pair encryption vs hiding that nuance. It does seem simpler that way but seems like something we'd want to run by Charles since it differs from current key pair usage.

Reply:
Assuming Charles is ok with hiding the private / key pair behind a "remember me" feature, I guess we could change the workflow to something like:

1. Prompt for username (remember last used)
2. If private key exists for selected user, use it. If not, prompt for password
3. Somewhere along the way present option to "remember me" and generate keys if not exists, or delete them if they do and user opts out

Reply:
Also, if we didn't want to make such a change on this PR, I could at least default the keygen username to remember the last username. That was just an oversight on my part

Reply:
I fixed it to remember the username